### PR TITLE
Expose configuration for handling coded values like -666666666

### LINKED ISSUE #72: API returns value of -666666666
Some calls to `acs5.state_county_tract()` return the value `-666666666`. This seems like a semantically significant value, but I can't find documentation on what it means.

For a reproducible example, this call returns the value in question:

```python
>>> c = Census(os.environ['CENSUS_API_KEY'], year=2016)
>>> c.acs5.state_county_tract('B19081_001E', 42, 101, '989100')

[{'B19081_001E': -666666666.0, 'state': '42', 'county': '101', 'tract': '989100'}]
```

Any thoughts on what this response means, and how I should handle it? So far I've just been converting them to nulls.

### PR DESCRIPTION
**Pull request by jeancochrane**
## Summary

Adjust the `Client.query` method to hande [coded values](https://www.census.gov/data/developers/data-sets/acs-1year/notes-on-acs-estimate-and-annotation-values.html), allowing the user to either:

1. Cast them to null
2. Raise an error

Set option 1) as the default. Implement these choices for the value `-666666666`, the only one I've encountered in the wild so far.

Closes #72.

## Testing instructions

- Confirm that all tests pass on Travis.

### CODE REVIEW

File: census/core.py [hunk 1]
```diff
@@ -36,7 +36,42 @@ def list_or_str(v):
...
+        # an estimate. See:
+        # https://www.census.gov/data/developers/data-sets/acs-1year/notes-on-acs-estimate-and-annotation-values.html
+        if str(v) == '-666666666':
+            if cast_nulls is True:
+                return None
+            else:
+                raise NullValueException('Unhandled coded value: ', str(v))
+        else:
+            return func(v)
+    return null_wrapper
```
Comment (jeancochrane) [inline census/core.py]: Another (maybe simpler) way to do this would be to avoid the decorator abstraction and use a "checking" function instead. I liked the decorator since it felt to me like a semantic way of indicating that this checking logic is a high-level modification that can be applied somewhat abstractly to the other type casting functions.

File: census/core.py [hunk 2]
```diff
@@ -158,6 +197,10 @@ def get(self, fields, geo, year=None, **kwargs):
 
     @retry_on_transient_error
     def query(self, fields, geo, year=None, **kwargs):
+        cast_nulls = kwargs.get('cast_nulls', True)
+        if cast_nulls not in [True, False]:
+            raise CensusException('cast_nulls argument must be True or False')
```
Comment (jeancochrane) [inline census/core.py]: I'm not totally confident in `cast_nulls` as the best name for the configuration argument that adjusts this behavior. In particular, it doesn't communicate that the alternative is to raise an error. Still, it was the best I had.

File: census/core.py [hunk 3]
```diff
@@ -187,10 +230,31 @@ def query(self, fields, geo, year=None, **kwargs):
...
+                for header, cast, item in zip(headers, types, d):
+                    if item is not None:
+                        try:
+                            result = {header: cast(item, cast_nulls)}
+                        except NullValueException:
+                            # This value needs to raise an error, but we need the
+                            # rest of the row values for context, so flag the
+                            # error and continue the iteration
+                            error = True
+                            result = {header: item}
```
Comment (jeancochrane) [inline census/core.py]: Some additional context for why we need to finish this inner loop before raising the error is that the elements generated by `for d in data` look something like this:

```python
{'var_name': -666666666, 'state': 42, 'county': 11}
```

The inner loop `for header, cast, item in zip(headers, types, d)` processes each of these elements in turn, returning a formatted "row" for the final results. The problem with raising an error in the inner loop is that we'll only know one of these `header: item` pairs (like `var_name: -666666666` or `'state': 42`) but unless you know all of the elements it's impossible to determine which API call raised the error.

File: census/tests/test_census.py [hunk 4]
```diff
@@ -121,6 +121,67 @@ def test_la_canada_2015(self):
...
+    """
+    Tests for handling coded values, like -666666666 and -999999999.
+    """
+    def test_handle_666666666(self):
+        """
+        Test the default behavior of handling -666666666 values, which is to
+        cast them to null.
+        """
+        # This call should return a value of -666666666
+        return_val = self._client.acs5.state_county_tract('B19081_001E',
```
Comment (fgregg) [inline census/tests/test_census.py]: Do we want this to be the default behavior? I think the default should be the behavior of cast_null=True, but I'm persuadable.

File: census/tests/test_census.py [hunk 5]
```diff
@@ -121,6 +121,67 @@ def test_la_canada_2015(self):
...
+                                                              101,
+                                                              '989100',
+                                                              year=2016,
+                                                              cast_nulls=False)
+
+    def test_handle_666666666_as_null(self):
+        """
+        Test casting -666666666 values to null.
+        """
+        return_val = self._client.acs5.state_county_tract('B19081_001E',
```
Comment (fgregg) [inline census/tests/test_census.py]: We should also test that a warning is emitted that a value was cast to a null. Particularly if we decide to make this the default behavior.

File: census/core.py (hunk 1 above)
Comment (fgregg) [reply census/core.py]: I think I'd be a bit more comfortable with this architecture if we had more cases than just the `-666666666`. That said I'm comfortable to go with this approach until we accumulate more cases, and then we can potentially refactor if appropriate.

File: census/tests/test_census.py [hunk 6]
```diff
@@ -121,6 +121,67 @@ def test_la_canada_2015(self):
...
+                                                          42,
+                                                          101,
+                                                          '989100',
+                                                          year=2016,
+                                                          cast_nulls=True)
+        self.assertEqual(return_val[0]['B19081_001E'], None)
+
+    def test_bad_cast_nulls_argument(self):
```
Comment (fgregg) [inline census/tests/test_census.py]: great to have these tests!
